enhancements/network/cloud-egress-ip.md

```diff
@@ -0,0 +1,445 @@
+  assignments to cloud VMs, in a cloud provider agnostic way.
+
+[#Summary]: #Summary
+
+### Non-Goals
+
+For now:
+
+- Assignment of external IP addresses to cloud VMs
+
+- Assignment of additional NICs to cloud VMs
+
+- Assignment of additional private IP address to any NIC besides the one the
+  network plugin manages.
+
+These three "Non-goals" could become "Goals" if we decide to do a generic "Cloud
+API operator" and write an enhancement proposal for that.  
+
+- Providing a cloud provider solution to OpenShift's networking feature: egress
+  router
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+#### Cloud provider data model for additional private IP assignment
+
+The following section describes initial researched information w.r.t. the API
+for private IP address assignment to VMs, on the three major clouds (AWS, GCP
+and Azure). The goal of this document is not to delve into the details of this,
+as the implementation will have to consider all the specifics. Instead it
+presents the general concepts as to give the reader an understanding of what
+will be required by this future component. The API for doing additional IP
+address assignment to VMs differ slightly between the clouds, but remain quite
+similar in the grand scheme of things.  
+
+##### Azure
+
+Azure requirement for assigning additional private IPs to an already existing
+NIC is detailed in [6]. Although the go SDK for communicating with the Azure API
+has not been studied yet, it is expected to resemble the Azure CLI API:
+
+```
+az network nic ip-config create \
+--resource-group $RG_NAME \
+--nic-name $NIC_NAME \
+--private-ip-address 10.0.0.6 \
+--name $name
+```
+
+Where `$RG_NAME` and `$NIC_NAME` needs to be discovered beforehand when
+assigning an additional IP. It is thus expected that this component will be able
+to retrieve such information prior to performing the IP assignment request to
+the Azure API. Note: the Azure doc does not make any particular reference to
+IPv4 / IPv6, so it's not clear whether both are supported. Azure does however
+claim to be a "dual-stack" cloud, so it is assumed that both are supported.
+
+[6]:
+https://docs.microsoft.com/en-us/azure/virtual-network/virtual-network-multiple-ip-addresses-cli#add
+
+##### AWS
+
+The AWS API in [7] discusses additional private IP assignment to existing NICs.
+AWS seems to differentiate their API between [IPv4]:
+
+```
+aws ec2 assign-private-ip-addresses --network-interface-id $NIC_ID --private-ip-addresses 10.0.0.82
+```
+
+and [IPv6] assignments:
+
+```
+aws ec2 assign-ipv6-addresses --network-interface-id $NIC_ID --ipv6-addresses 2001:db8:1234:1a00:3304:8879:34cf:4071
+
+```
+
+[7]:
+https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/MultipleIP.html#ManageMultipleIP
+[IPv4]:
+https://docs.aws.amazon.com/cli/latest/reference/ec2/assign-private-ip-addresses.html
+[IPv6]:
+https://docs.aws.amazon.com/cli/latest/reference/ec2/assign-ipv6-addresses.html
```

Comment:
According to the documentation, assigning IP addresses can be asynchronous (in the event the IP is moved from another node, which we would do). The node itself needs to read from the instance metadata endpoint and report IPs that are assigned.

We may not care about this, but something to think about.

Reply:
I will mention this